File: frontend/mobile/lib/pages/Timer/timer_utils.dart
import 'package:flutter/material.dart';

import '../../common/DateTimeFormat.dart';

const studyTypeLabels = <String, String>{
  'reading': 'อ่านตำรา/เอกสาร',
  'practice': 'ทำโจทย์/ฝึกปฏิบัติ',
  'video': 'ดูวิดีโอ/lecture',
  'review': 'ทบทวน/สรุปบทเรียน',
};

const studyTypeColors = <String, Color>{
  'reading': Color(0xFF91C9EF),
  'practice': Color(0xFFF6B7CC),
  'video': Color(0xFFF5C779),
  'review': Color(0xFF9ED7BD),
};

String formatClock(int totalSeconds) {
  final safeSeconds = totalSeconds.clamp(0, 359999);
  final hours = safeSeconds ~/ 3600;
  final minutes = (safeSeconds % 3600) ~/ 60;
  final seconds = safeSeconds % 60;
  return [
    hours,
    minutes,
    seconds,
  ].map((value) => value.toString().padLeft(2, '0')).join(':');
}

String formatStudyDuration(double minutes, {bool compact = false}) {
  final safeMinutes = minutes.round().clamp(0, 999999);
  final hours = safeMinutes ~/ 60;
  final remainingMinutes = safeMinutes % 60;
  if (compact) {
    return '${hours.toString().padLeft(2, '0')} ชม. '
        '${remainingMinutes.toString().padLeft(2, '0')} นาที';
  }
  if (hours > 0) return '$hours ชั่วโมง $remainingMinutes นาที';
  return '$remainingMinutes นาที';
}

String formatThaiMonth(String monthKey) {
  final parts = monthKey.split('-');
  final year = parts.isNotEmpty ? int.tryParse(parts[0]) : null;
  final month = parts.length > 1 ? int.tryParse(parts[1]) : null;
  if (year == null || month == null || month < 1 || month > 12) {
    return monthKey;
  }
  return formatDisplayMonthYear(year, month);
}

String formatThaiDateTime(DateTime? value) {
  if (value == null) return 'ไม่ทราบเวลา';
  return formatDisplayDateTime(value);
}

File: frontend/web/user/src/components/Timer/timer.utils.ts
import type { StudyTypeName } from "@/interfaces/time.interface";
import { formatDisplayMonthYear } from "@/utils/dateTime";

export const studyTypeLabels: Record<StudyTypeName, string> = {
  reading: "อ่านตำรา/เอกสาร",
  practice: "ทำโจทย์/ฝึกปฏิบัติ",
  video: "ดูวิดีโอ/lecture",
  review: "ทบทวน/สรุปบทเรียน",
};

export const studyTypeColors: Record<StudyTypeName, string> = {
  reading: "#91c9ef",
  practice: "#f6b7cc",
  video: "#f5c779",
  review: "#9ed7bd",
};

export const formatClock = (totalSeconds: number) => {
  const safeSeconds = Math.max(0, Math.floor(totalSeconds));
  const hours = Math.floor(safeSeconds / 3600);
  const minutes = Math.floor((safeSeconds % 3600) / 60);
  const seconds = safeSeconds % 60;
  return [hours, minutes, seconds]
    .map((value) => String(value).padStart(2, "0"))
    .join(":");
};

export const formatDuration = (minutes: number, compact = false) => {
  const safeMinutes = Math.max(0, Math.round(minutes));
  const hours = Math.floor(safeMinutes / 60);
  const remainingMinutes = safeMinutes % 60;
  if (compact) {
    return `${String(hours).padStart(2, "0")} ชม. ${String(
      remainingMinutes
    ).padStart(2, "0")} นาที`;
  }
  return hours > 0
    ? `${hours} ชั่วโมง ${remainingMinutes} นาที`
    : `${remainingMinutes} นาที`;
};

export const formatThaiMonth = (monthKey: string) => {
  return formatDisplayMonthYear(monthKey);
};
